docs/style-guide.md

```diff
@@ -0,0 +1,439 @@
+
+### Don't have "mixed" Crates :: suggested
+
+The current thinking is that you should either:
+
+* Use a single `lib.rs` file for a Crate with everything in it
+* Use multiple Rust sources and then only use `lib.rs` to export modules
+
+Which of these options you chose is dependent on the size/complexity of the module(s).
+
+There might be cases where an intermediate "mixed" Crate is worthwhile, but it should be clearly documented as to why it is necessary.
+
+## Shared Internal Code
+
+Rust supports having [non-public sub-modules](https://doc.rust-lang.org/reference/visibility-and-privacy.html) within a Crate, and only the top-level "implicit" module is exposed by default (other modules can be exported by the top level module). See also [File hierarchy - Rust By Example](https://doc.rust-lang.org/rust-by-example/mod/split.html).
+
+### Avoid over-exposing internal APIs :: suggested
+
+When sharing an API between modules in a Crate, use `pub(<scope>)` to express visibility, and select the smallest scope which suits the intent of the API (i.e. prefer `pub(crate::specific_mod)` to `pub(crate)` where appropriate).
+
+# Structs and Traits
+
+For ICU4X we should be looking to keep code consistent and very unsurprising, especially for layout and behavior of data structures.
+
+## Private vs Public
+
+Rust offers a compelling model for encapsulating implementation details based on a hierarchical view of privacy; see [Visibility and Privacy - The Rust Reference](https://doc.rust-lang.org/reference/visibility-and-privacy.html). This should allow ICU4X to encapsulate all of its implementation details neatly from applications using the library and avoid too many cases of [Hyrum's Law](https://www.hyrumslaw.com/).
+
+### Minimize user-visible structs :: suggested
+
+While this sounds a bit obvious, I think it's important to stress that with Rust's privacy model, there is never any need to have an internal type be user-visible, and once a type is user-visible outside a Crate, it's published and should not be changed.
+
+The only place this might be an issue is if we decide to deliver ICU4X as multiple Crates. If we do that then we just added a requirement that our inter-Crate APIs be public (there's no privileged sharing outside Crates). This is one reason I propose we deliver ICU4X as a single Crate with multiple modules.
+
+There is a known pattern of using [`doc(hidden)`](https://doc.rust-lang.org/rustdoc/the-doc-attribute.html#dochidden) and having module names prefixed with multiple underscores to indicate they must not be relied upon outside the project, but ideally APIs would be deesignedto minimize the need for this and the expected external usage would be clearly documented for future maintainers.
```

Comment:
I disagree that we should *only* release components as part of a single crate. Creates like `icu-locale` are useful without the whole ecosystem and I'd like to allow people to use just that. That's the `modular` in our description.

So I'd prefer to say that we want to pay special attention to minimize the public API surface, and we recognize three areas which may warrant a technically public API that we don't consider stable and will mark with `doc(hidden)`:
 - Where another crate may need access to non-public API
 - Where we want to expose some non-public API for tests
 - Where we want to expose some non-public API for helper macros

Reply:
Also, spelling -> `deesignedto`

Reply:
Done. And thanks for the clarification.